common/network-types/src/udp.rs

```diff
@@ -1,152 +1,508 @@
-/// 1) setting it via [`with_counterparty`](ConnectedUdpStream::with_counterparty) after binding
+/// 1) setting it during build via [`UdpStreamBuilder::with_counterparty`]
 /// 2) receiving some data from the other side.
 ///
 /// Whatever of the above happens first, sets the counterparty.
 /// Once the counterparty is set, all data sent and received will be sent or filtered by this
 /// counterparty address.
 ///
 /// If data from another party is received, an error is raised, unless the object has been constructed
-/// with [`with_foreign_data_discarding`](ConnectedUdpStream::with_foreign_data_discarding) setting.
-#[derive(Debug)]
+/// with [`ForeignDataMode::Discard`] or [`ForeignDataMode::Accept`] setting.
+///
+/// This object is also capable of parallel processing on a UDP socket.
+/// If [parallelism](UdpStreamBuilder::with_parallelism) is set, the instance will create
+/// multiple sockets with `SO_REUSEADDR` and spin parallel tasks that will coordinate data and
+/// transmission retrieval using these sockets. This is driven by RX/TX MPMC queues, which are
+/// per-default unbounded (see [queue size](UdpStreamBuilder::with_queue_size) for details).
 pub struct ConnectedUdpStream {
-    sock: UdpSocket,
-    counterparty: Option<std::net::SocketAddr>,
-    closed: bool,
-    foreign_data_mode: ForeignDataMode,
+    socket_handles: Vec<(tokio::task::JoinHandle<()>, tokio::task::JoinHandle<()>)>,
+    ingress_rx: Box<dyn AsyncRead + Send + Unpin>,
+    egress_tx: Option<BoxIoSink<Box<[u8]>>>,
+    counterparty: Arc<OnceLock<std::net::SocketAddr>>,
+    bound_to: std::net::SocketAddr,
 }
 
 /// Defines what happens when data from another [`SocketAddr`](std::net::SocketAddr) arrives
 /// into the [`ConnectedUdpStream`] (other than the one that is considered a counterparty for that
 /// instance).
-#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Default)]
 pub enum ForeignDataMode {
     /// Foreign data are simply discarded.
     Discard,
     /// Foreign data are accepted as if they arrived from the set counterparty.
     Accept,
     /// Error is raised on the `poll_read` attempt.
+    #[default]
     Error,
 }
 
-impl ConnectedUdpStream {
-    /// Binds the UDP socket to the given address, without an assigned counterparty.
-    pub async fn bind<A: tokio::net::ToSocketAddrs>(bind: A) -> tokio::io::Result<Self> {
-        Ok(Self {
-            sock: UdpSocket::bind(bind).await?,
-            counterparty: None,
-            closed: false,
-            foreign_data_mode: ForeignDataMode::Error,
-        })
+/// Builder object for the [`ConnectedUdpStream`].
+///
+/// If you wish to use defaults, do `UdpStreamBuilder::default().build(addr)`.
+#[derive(Debug, Default)]
+pub struct UdpStreamBuilder {
+    foreign_data_mode: Option<ForeignDataMode>,
+    buffer_size: Option<usize>,
+    queue_size: Option<usize>,
+    parallelism: Option<usize>,
+    counterparty: Option<std::net::SocketAddr>,
```

Comment:
Why all optional? at least the parallelism and sizes should have a reasonable default.

Reply:
Some of them can be made non-optional.